### ISSUE TITLE
arangodb apt-get upgrade fails

### ISSUE THREAD

**Issue by tomwolf** [arangodb/arangodb#2452]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [ ] 3.0
- [x ] 3.1
- [ ] 3.2 pre-releases: ___
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x ] Linux 
  - [ x] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [ ] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an installation-related issue:
apt-get upgrade fails with following error:

Failed to perform requested operation on package.  Trying to recover:
Setting up arangodb3 (3.1.18) ...
Upgrading database files.
2017-04-25T10:33:26Z [31016] FATAL database is locked, please check the lock file '/var/lib/arangodb3/LOCK'
dpkg: error processing package arangodb3 (--configure):
 subprocess installed post-installation script returned error exit status 1
Errors were encountered while processing:
 arangodb3

It looks like that the installation does not stop arangodb before upgrading.

**Comment by kvahed** [comment]
Is your database service still running? What does `ps -eaf|grep arangod` reveal?
You should stop the database server for version upgrades.
`sudo service arangodb3 stop`
upgrade and restart the server afterwards
`sudo service arangodb3 start`

**Comment by dothebart** [comment]
In some cases the old arangodb process fails to shut down. You need to stop the old process manually using `killall arangod` and once its gode re-start it using `/etc/init.d/arangodb start` 

We haven't been able to insulate that incident yet; do you have some interesting output in the logfile of the supervisor?

**Comment by tomwolf** [comment]
It happened at 3 servers today.
For debugging reason I downgraded a machine to 3.1.17 to reproduce the error but everything runs fine.
Excerpt from arangod.log during failed upgrade :
<pre>
2017-04-25T10:33:26Z [31009] WARNING fcntl on lockfile '/var/lib/arangodb3/LOCK' failed: not implemented. a possible reason is that the filesystem
 does not support file-locking
2017-04-25T10:33:26Z [31009] FATAL database is locked, please check the lock file '/var/lib/arangodb3/LOCK'
2017-04-25T10:33:26Z [31016] FATAL database is locked, please check the lock file '/var/lib/arangodb3/LOCK'
</pre>

Sorry, maybe we should close the ticket...

**Comment by dothebart** [comment]
yea, there should also be an `arangodb.log.supervisor` which will contain more information with 3.1.18 in this case (hopefully). 
So maybe we can track down that situation on next update. 
Its a bit anoying that we can't easily reproduce this problem - but hopefully soon ;-)

Ok, lets close it. If you get more information feel free to comment again.

**Comment by bzkdjc** [comment]
This Problem has persisted here since I installed ArangoDB for the first time _(v3.1.16)_ on my server with **Ubuntu 16.10** and **BtrFS**. 
Problem was present at each upgrade from 3.1.16 till 3.1.20! 

So, as I cannot (again) upgrade from 3.1.19 to 3.1.20, I checked **arangod.log.supervisor**  and just saw: 
```
2017-04-06T04:57:09Z [4653] INFO {startup} received SIGINT for supervisor
2017-04-06T12:59:52Z [10615] INFO {startup} received SIGINT for supervisor
2017-04-24T23:37:21Z [30120] INFO {startup} supervisor has forked a child process with pid 30121
2017-04-24T23:41:16Z [30120] INFO {startup} received SIGINT for supervisor; commanding client [30121] to shut down.
2017-04-24T23:41:20Z [30120] INFO {startup} waitpid woke up with return value 30121 and status 0 and DONE = true
2017-04-24T23:43:34Z [4689] INFO {startup} supervisor has forked a child process with pid 4690
2017-05-03T09:29:48Z [11962] INFO {startup} supervisor has forked a child process with pid 11963
2017-05-03T09:31:23Z [11962] INFO {startup} received SIGINT for supervisor; commanding client [11963] to shut down.
2017-05-03T09:31:27Z [11962] INFO {startup} waitpid woke up with return value 11963 and status 0 and DONE = true
2017-05-03T09:33:28Z [3610] INFO {startup} supervisor has forked a child process with pid 3611
``` 

But **arangod.log** is better and  shows:
```
2017-05-18T16:29:08Z [580] INFO ArangoDB 3.1.19 [linux] 64bit, using VPack 0.1.30, ICU 54.1, V8 5.0.71.39, OpenSSL 1.0.2g  1 Mar 2016
2017-05-18T16:29:08Z [580] INFO file-descriptors (nofiles) hard limit is 800000, soft limit is 131072
2017-05-18T16:29:08Z [580] WARNING fcntl on lockfile '/var/lib/arangodb3/LOCK' failed: not implemented. a possible reason is that the filesystem does not support file-locking
2017-05-18T16:29:08Z [580] FATAL database is locked, please check the lock file '/var/lib/arangodb3/LOCK'
```
which seems to be more detailed! 

Does this mean that BtrFS is the cause of the problem? 
Should BtrFS not be used with ArangoDB?

**Comment by dothebart** [comment]
Hi, 
we located & fixed the error with ArangoDB 3.1.20. It took us a while to figure out what was happening since we only saw the Supervisor being gone **after** the upgrade; but in fact it was gone **after one day of operation**.

The process setup looks like that: 
`Supervisor(as root) --forks-->Databaseserver`
The Supervisor is intended to remain up and running. However, the supervisor needs to handle a special situation: the logrotation. Therefore `logrotate.d` renames the currently open logfile, sends the Supervisor a `SIG_HUP` so the DBServer re-opens its logfile on the original filename. This has been implemented inside of ArangoDB for quiet a while since its standard unix behaviour to act like this.

However, the final recipient for this signal should be the Database server - not the Supervisor. 
So up to 3.1.19 the Supervisor would not react on HUP, so the default behaviour is to terminate the process - which is why it goes away and the DBServer remains running. 

Once this happenes the init script doesn't know how to reach the DBServer anymore - thus you have to terminate the DBServer manually. The upgrade to 3.1.20 should be the last time this is neccessary.

Regarding BTRFS - we discourage using it - see https://docs.arangodb.com/devel/Manual/Administration/ It tends to give performance penalties and the way the MMFiles engine operates it isn't handled in a good way. So yes, you should prefer EXT4.
